docs/src/lens.md

```diff
@@ -1,149 +1,251 @@
+
+```rust
+#[derive(Lens)]
+struct Outer {
+    inner: Inner,
+}
 
-Let's build the UI for our todo list item:
+#[derive(Lens)]
+struct Inner {
+    text: String
+}
 
-```rust,noplaypen
-{{#include ../book_examples/src/lens_md.rs:build_ui}}
+// `composed_lens` will contain a lens that goes from `Outer` through `Inner` to `text`.
+let composed_lens = Outer::inner.then(Inner::text);
 ```
 
-## Advanced lenses
+> Side note: Because unlike Haskell, Rust has all the type information during monomorphisation, it can inline all these functions and the extra cost of having 2 lens functions disappears, leaving what you would have written by hand. When you're waiting for Rust's infamously long compilations to finish, know that you're waiting for a potentially significant run-time benefit in exchange.
 
-Field access is a very simple (and common, and *useful*) case, but lenses can do much more than that.
+`LensExt` contains a few more useful methods for handling things like negating a boolean, or auto-`Deref`ing a value.
 
-### `LensExt` and combinators
+There are also 3 special structs in `druid::lens`: `Constant`, `Identity` and `Unit`. `Constant` is a lens that always returns the same value, and always discards any changes, while `Identity` is a lens that does nothing. You might say "what is the point of a lens that does nothing", which would be a fair question. Well, there are some places where a lens is required, and having an identity allows the user to say act as if there was no lens. It's also used to begin a composition chain using the combinators like `then`. `Unit` is a special case of `Constant` where the constant in question is `()`.
 
-Similar to the `WidgetExt` trait, we offer a `LensExt` trait that provides
-various functions for composing lenses. These are similar to the various methods
-on iterator; you can `map` from one lens to another, you can index into a
-collection, or you can efficiently access data in an `Arc` without unnecessary
-mutation; see the main crate documentation for more.
+> Side note: Because `()` only has 1 value, it does actually respect mutations. It's just that mutations always result in the same value again (`()`).
 
-As your application gets more complicated, it will become likely that you want
-to use fancier sorts of lensing, and `map` and company can start to get out
-of hand; when that happens, you can always implement a lens by hand.
+### The `lens` macro
 
-### Getting something from a collection
+Finally, there is a macro for constructing lenses on the fly. It allows you to lens into fields of a struct you don't control (so you can't derive `Lens` for it), it also allows lensing into tuples and tuple structs, and lastly it will create index lenses into slices.
```

Comment:
Definitely good to bring this up. I think we need to demonstrate how this macro invocation is written, though. Maybe a contrived example on a standard library type? (Length of a vec?)